File: Step03_Arrays/Lec1_Easy/Program_06_optimal.cpp
#include <bits/stdc++.h>
using namespace std;

// Left Rotate an array by 'd' places

int main() {
    int n, d;
    cin >> n;
    cin >> d;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    d = d % n;
    auto it = arr.begin();
    reverse(it, it + d);      // O(d)
    reverse(it + d, it + n);  // O(n-d)
    reverse(it, it + n);      // O(n)

    for (auto element : arr) {
        cout << element << " ";
    }
}

// Time-complexity  : O(2n)
// Space-complexity : O(1)

File: Step03_Arrays/Lec1_Easy/Program_06_optimal.java
package Step03_Arrays.Lec1_Easy;

import java.util.*;

// Left Rotate an array by 'd' places

public class Program_06_optimal {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int n = in.nextInt();
        int d = in.nextInt();
        ArrayList<Integer> arr = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            arr.add(in.nextInt());
        }

        d = d % n;

        // O(d)
        Collections.reverse(arr.subList(0, d));

        // O(n-d)
        Collections.reverse(arr.subList(d, n));

        // O(n)
        Collections.reverse(arr.subList(0, n));

        for (var element : arr) {
            System.out.print(element + " ");
        }

        in.close();
    }
}

// Time-complexity : O(2n)
// Space-complexity : O(1)
